staging/src/k8s.io/api/coordination/v1alpha1/types.go

```diff
@@ -0,0 +1,77 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +k8s:prerelease-lifecycle-gen:introduced=1.31
+
+// LeaseCandidate defines a candidate for a lease object.
+// Candidates are created such that coordinated leader election will pick the best leader from the list of candidates.
+type LeaseCandidate struct {
+	metav1.TypeMeta `json:",inline"`
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+	// spec contains the specification of the Lease.
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#spec-and-status
+	// +optional
+	Spec LeaseCandidateSpec `json:"spec,omitempty" protobuf:"bytes,2,opt,name=spec"`
+}
+
+// LeaseSpec is a specification of a Lease.
+type LeaseCandidateSpec struct {
+	// BinaryVersion is the binary version
+	BinaryVersion string `json:"binaryVersion,omitempty" protobuf:"bytes,5,opt,name=binaryVersion"`
```

Comment:
a related question, is there some need of validation between `binVersion` and `compatVersion` as:
`binVersion` must be greater or equal to `compatVersion` ?

Reply:
talked to Siyuan around compatibilityVersion and I do believe we should not use two versions and just use the emulated version. Furthermore, min compatibility version is a storage/CEL concept and does not apply to components like kcm, kube-scheduler, etc.

I'll replace compatVersion with emulatedVersion in your example. In this case,
1: binVersion=1.32, emulatedVersion=1.30
2: binVersion=1.31, emulatedVersion=1.31

we should always pick the lowest emulated version and the correct logic should be incorrect. In cases with the same emulatedVersion and different binary versions, I don't really think it matters too much which one is leader since features introduced after the emulated version are disabled anyway. Furthermore, with the introduction of emulatedVersion, version skew is defined by the emulatedVersion and binaryVersion is disregarded.

I propose that we should only use one field `emulatedVersion` and pick the minimum version. Unfortunately emulatedVersion is not a semver and is a <major>.<minor> [format](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/util/version/version.go#L92), but we can still write proper validation for that field.

/cc @jpbetz @sttts

Reply:
Yes, this matches my understanding-- emulationVersion is really what matters when it is available.  If we happen to have two candidates wit hteh same compatVersion and different binary versions, then having a consistent rule for which is selected is useful, but that's more of a tie breaker situation, not the core rule.

Reply:
sol lowest emulation version, if those match is it lowest or highest binary version?  The devil you know or the devil you don't ? :)

Reply:
If two candidates are on the same emulationVersion, do we really care about binary version? Sure it's slightly more accurate to tiebreak on it but the goal of this KEP is to satisfy kubernetes version skew. The compatibility version KEP rewrites kubernetes version skew policy to be only based on emulated version. I think it makes sense to only care about the emulated version when selecting leader in coordinated leader election? We could tiebreak with something like creationTimestamp and remove binary version from consideration.

Reply:
Newer binaries SHOULD have old bugs fixed, but may have new bugs.

Deterministic > non-deterministic.  That's my concern.  Joe may feel more strongly.

Reply:
> Deterministic > non-deterministic. That's my concern. Joe may feel more strongly.

We anticipate that compatibility version will be used to enable cluster admins to upgrade the binary version while pinning the compatibility version. When this is done, the only difference between leader elected components will be the binary version.  Having deterministic/predictable binary version selection during such upgrades is something I very much want.

It's not expensive to add is it?

Reply:
Not expensive, just wanted to better understand the use case. I have a slight preference for older binary version so the logic looks more consistent by saying "Oldest" strategy (first pick oldest ev, then pick oldest bv). 

Are clusters able to upgrade remaining on the same patch version? Do we need to consider things like upgrading from 1.31.0-mycloudpatch.1 to 1.31.0-mycloudpatch.2? I'm curious if we can strongly guarantee the determinism around binary version or if it can only be done on a best effort basis.

Reply:
WRT version format -- is the goal to make this functionality very specific to kubernetes and our sort of version numbers? Or is this a general purpose API, in which case we should define it in a more public and general purpose way. Even if that means that for kubernetes we convert our version into a proper semver or something.

Reply:
Isn't having two versions (binary + emulated) already quite kube opinionated? I'd imagine anything outside of kube will only care about one or set them to be equal. I do like semver and having easily validated fields though.

Reply:
> Isn't having two versions (binary + emulated) a bit kube opinionated already?

A little, but not as much as you might think.  Go does the same, for example.  

> Since emulation version is only major.minor, patch version will always be 0

Are you saying that you will accept "x.y" and convert it to "x.y.0" or that you will demand "x.y.z" as input, and the client must set the trailing .0 ?  I think either can work as long as we spec it clearly

Reply:
> Are you saying that you will accept "x.y" and convert it to "x.y.0" or that you will demand "x.y.z" as input, and the client must set the trailing .0 ? I think either can work as long as we spec it clearly

Leaning towards the latter. Accept and validate only semver. The client is responsible for providing a valid semver.